json/src/fs2/data/json/selector/dsl.scala

```diff
@@ -0,0 +1,175 @@
+}
+
+object IteratorBuilder {
+
+  implicit def LenientableIterator: Lenientable.Aux[IteratorBuilder[Strict], IteratorBuilder[Lenient], NotApplicable] =
+    new Lenientable[IteratorBuilder[Strict], NotApplicable] {
+    type Out = IteratorBuilder[Lenient]
+    def makeLenient(builder: IteratorBuilder[Strict]): IteratorBuilder[Lenient] =
+      builder.copy(strict = false)
+  }
+
+}
+
+case class IndicesBuilder[S](predicate: IndexPredicate, strict: Boolean, parent: SelectorBuilder[_, _])
+    extends SelectorBuilder[NotApplicable, S] {
+  def compile: Selector = Selector.PipeSelector.from(parent.compile, Selector.IndexSelector(predicate, strict))
+}
+
+object IndicesBuilder {
+
+  implicit def LenientableIndices: Lenientable.Aux[IndicesBuilder[Strict], IndicesBuilder[Lenient], NotApplicable] =
+    new Lenientable[IndicesBuilder[Strict], NotApplicable] {
+      type Out = IndicesBuilder[Lenient]
+      def makeLenient(builder: IndicesBuilder[Strict]): IndicesBuilder[Lenient] =
+        builder.copy(strict = false)
+    }
+
+}
+
+case class NamesBuilder[M, S](predicate: NamePredicate,
+                              strict: Boolean,
+                              mandatory: Boolean,
+                              parent: SelectorBuilder[_, _])
+    extends SelectorBuilder[M, S] {
+  def compile: Selector =
+    Selector.PipeSelector.from(parent.compile, Selector.NameSelector(predicate, strict, mandatory))
+}
+
+object NamesBuilder {
+
+  implicit def LenientableNames[M]: Lenientable.Aux[NamesBuilder[M, Strict], NamesBuilder[M, Lenient], M] =
+    new Lenientable[NamesBuilder[M, Strict], M] {
+      type Out = NamesBuilder[M, Lenient]
+      def makeLenient(builder: NamesBuilder[M, Strict]): NamesBuilder[M, Lenient] =
+        builder.copy(strict = false)
+    }
+
+  implicit def MandatoriableNames[S]: Mandatoriable.Aux[NamesBuilder[Optional, S], NamesBuilder[Mandatory, S], S] =
+    new Mandatoriable[NamesBuilder[Optional, S], S] {
+      type Out = NamesBuilder[Mandatory, S]
+      def makeMandatory(builder: NamesBuilder[Optional, S]): NamesBuilder[Mandatory, S] =
+        builder.copy(mandatory = true)
+    }
+
+}
+
+class NotApplicable private {}
+class Optional private {}
+class Mandatory private {}
+class Strict private {}
+class Lenient private {}
+
+sealed trait Lenientable[In, M] {
+  type Out <: SelectorBuilder[M, Lenient]
+  def makeLenient(builder: In): Out
+}
+
+object Lenientable {
+  @implicitNotFound(msg = "There seems to be no way to make this selector lenient. Is it already?")
+  type Aux[In, Out0, M] = Lenientable[In, M] { type Out = Out0 }
+}
+
+sealed trait Mandatoriable[In, S] {
+  type Out <: SelectorBuilder[Mandatory, S]
+  def makeMandatory(builder: In): Out
+}
+
+object Mandatoriable {
+  @implicitNotFound(msg = "Only not yet mandatory field selectors can be made mandatory")
+  type Aux[In, Out0, S] = Mandatoriable[In, S] { type Out = Out0 }
+}
+
+object SelectorBuilder {
```

Comment:
Very clever encoding, I like it a lot!

Just for the readability of the source code, I would push this companion object up, next to the `SelectorBuilder` class.

Reply:
Thanks. And true.